File: src/main/java/com/github/zametki/db/sql/UsersSql.java
package com.github.zametki.db.sql;

import com.github.mjdbc.Bind;
import com.github.mjdbc.BindBean;
import com.github.mjdbc.Sql;
import com.github.zametki.model.User;
import com.github.zametki.model.UserId;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

/**
 * Set of 'users' table queries
 */
public interface UsersSql {

    @Nullable
    @Sql("SELECT * FROM users WHERE id = :id")
    User selectUserById(@Bind("id") UserId userId);

    @Nullable
    @Sql("SELECT * FROM users WHERE login = :login")
    User selectUserByLogin(@Bind("login") String login);

    @Nullable
    @Sql("SELECT * FROM users WHERE email = :email")
    User selectUserByEmail(@Bind("email") String email);

    @NotNull
    @Sql("INSERT INTO users (login, password_hash, email, uid, registration_date, termination_date, last_login_date, settings) " +
            "VALUES (:login, :passwordHash, :email, :uid, :registrationDate, :terminationDate, :lastLoginDate, :settings)")
    UserId insertUser(@BindBean User user);

    @Sql("UPDATE users SET password_hash = :hash WHERE id = :id")
    void updatePasswordHash(@Bind("id") UserId id, @Bind("hash") String passwordHash);

    @Sql("UPDATE users SET settings = :settings WHERE id = :id")
    void updateSettings(@BindBean User user);

    @Sql("UPDATE users SET last_login_date = :lastLoginDate WHERE id = :id")
    void updateLastLoginDate(@BindBean User user);
}

File: src/main/java/com/github/zametki/db/sql/UsersSql.kt
package com.github.zametki.db.sql

import com.github.mjdbc.Bind
import com.github.mjdbc.BindBean
import com.github.mjdbc.Sql
import com.github.zametki.model.User
import com.github.zametki.model.UserId

/**
 * Set of 'users' table queries
 */
interface UsersSql {

    @Sql("SELECT * FROM users WHERE id = :id")
    fun selectUserById(@Bind("id") userId: UserId): User?

    @Sql("SELECT * FROM users WHERE login = :login")
    fun selectUserByLogin(@Bind("login") login: String): User?

    @Sql("SELECT * FROM users WHERE email = :email")
    fun selectUserByEmail(@Bind("email") email: String): User?

    @Sql("INSERT INTO users (login, password_hash, email, uid, registration_date, termination_date, last_login_date, settings) " + "VALUES (:login, :passwordHash, :email, :uid, :registrationDate, :terminationDate, :lastLoginDate, :settings)")
    fun insertUser(@BindBean user: User): UserId

    @Sql("UPDATE users SET password_hash = :hash WHERE id = :id")
    fun updatePasswordHash(@Bind("id") id: UserId, @Bind("hash") passwordHash: String)

    @Sql("UPDATE users SET settings = :settings WHERE id = :id")
    fun updateSettings(@BindBean user: User)

    @Sql("UPDATE users SET last_login_date = :lastLoginDate WHERE id = :id")
    fun updateLastLoginDate(@BindBean user: User)
}
